### ISSUE TITLE
Eliminate SCSS variable duplication

### ISSUE THREAD

**Issue by shawnbot** [18F/web-design-standards#1535]
Currently, our "core" SCSS variables are duplicated in [_defaults.scss](https://github.com/18F/web-design-standards/blob/master/src/stylesheets/core/_defaults.scss) and [_variables.scss](https://github.com/18F/web-design-standards/blob/master/src/stylesheets/core/_variables.scss), and our grid settings are duplicated _again_ in [_grid-settings.scss](https://github.com/18F/web-design-standards/blob/master/src/stylesheets/core/_grid-settings.scss), as noted in #1532.

This is a problem: It's confusing for us to maintain, and it's confusing for users of the Sass source who want to customize the system. I've noted most of the documentation-related improvements on customization guidance in #1534, but this issue is about the code itself, and how we can make our SCSS variables easier for us to maintain and for our users to selectively override.

In [the docs](https://standards.usa.gov/getting-started/developers/#css-architecture), we specifically suggest that users copy `_variables.scss` and import the modified copy before the rest of the uswds source. The problem with this advice is that copying all of the variables locks that user into all of the default values at the time they copied that file. **The solution is to only set variables that you intend on overriding.** Mechanically, here's how it should work:
1. Remove [_defaults.scss](https://github.com/18F/web-design-standards/blob/master/src/stylesheets/core/_defaults.scss) (or deprecate them so as not to introduce issues for users, as I've done in #1530)
2. Add `!default` to all of our variable declarations in [_variables.scss](https://github.com/18F/web-design-standards/blob/master/src/stylesheets/core/_variables.scss) so that they can all be overridden
3. Update the [reference](https://github.com/18F/web-design-standards/blob/8f14a0c93ba6cb199b07d39e64a2dba4e0431a35/src/stylesheets/all.scss#L10) in our SCSS entrypoint from `core/defaults` to `core/variables`

Then, instead of maintaining a copy of `_variables.scss`, users would remove all of the variables they _haven't_ changed, which prevents them from being "locked in" to those values even if we change them in a later release.

**Comment by shawnbot** [comment]
Cc'ing @joshbruce on this. Josh: Am I on the right track here? Does this capture #1532, and can we consolidate the issues if so?

**Comment by shawnbot** [comment]
I'd like to merge #1530 before getting started on this, and work in parallel on release notes that document how users should update their Sass to accommodate the next release. My hope would be that we can get users to refactor their imports down to something like:

```scss
// customizations
@import 'variables';
// FKA 'all'
@import 'uswds/uswds';
```

or, if selectively importing different components:

```scss
@import 'variables';
@import 'uswds/variables';
@import 'uswds/core/typography';
// etc.
```

Of course, to do this, we would need to nest all of our partials under a  `uswds` directory within our `src/stylesheets` hierarchy, which is not what #1530 does yet, so... 😬

**Comment by shawnbot** [comment]
And if we're unsure what to do, we should probably take some time to learn more about how people are currently customizing the standards by [searching on github](https://github.com/search?q=uswds+variables.scss&ref=searchresults&type=Code&utf8=%E2%9C%93) and reaching out to those users.

**Comment by shawnbot** [comment]
Okay, so I did some digging around, and I found when searching github for instances of "uswds" + [variables](https://github.com/search?utf8=%E2%9C%93&q=uswds+core%2Fvariables&type=Code&ref=searchresults), [defaults](https://github.com/search?utf8=%E2%9C%93&q=uswds+core%2Fdefaults&type=Code&ref=searchresults), and [grid-settings](https://github.com/search?utf8=%E2%9C%93&q=uswds+core%2Fgrid-settings&type=Code&ref=searchresults) that there are a couple of different patterns of use for these three files, ranked in terms of popularity:

1. Import uswds' all.scss, _then_ modify variables:
  * [SAM WDS](https://github.com/GSA/sam-web-design-standards/blob/d2792eae0d4b8f4862d9785c378d85f1c369c9cf/src/stylesheets/all.scss#L10-L14)
  * [USDA-FSA](https://github.com/USDA-FSA/fsa-design-system/blob/gh-pages/_sass/_uswds.scss#L15-L18)
  * [USDA-FSA docs](https://github.com/USDA-FSA/fsa-design-system/blob/gh-pages/_sass/fsa-style-docs.scss#L7-L12)
  * [VA Healthcare Application](https://github.com/department-of-veterans-affairs/healthcare-application/blob/master/assets/css/main.scss#L35)
  * [vets.gov](https://github.com/department-of-veterans-affairs/vets-website/blob/4cac81f8a8ff4a9bbec15f2edac096c049cddd3f/src/sass/style.scss#L36)
  * [cloud.gov style (cg-style)](https://github.com/18F/cg-style/blob/529c5d3cb292cdd3d56b3c38d17de25accf6c2d0/src/css/_core.scss)
  * [this Jekyll site](https://github.com/carlosvalle/jekyll/blob/515ce8c0f2f715bbe24adf7f4fa10c0b43fb956f/src/stylesheets/all.scss#L10-L15)
  * [this "forked CFDA Angular UI"](https://github.com/anuragmundada-reisys/forkedCFDAAngularUI/blob/76004984aa2352c7ab89b81136db2961bef389cd/src/main/scss/uswds/_uswds-modified.scss#L11-L24)
1. Copy \_variables.scss and import that _instead of ours_, as we suggest in the docs:
  * [USAJOBS design system](https://github.com/USAJOBS/design-system/blob/c4f8b395219c701660a11aac922961fe7596bdd4/_scss/all-usajobs.scss#L10) has its own, heavily modified [version of _variables.scss](https://github.com/USAJOBS/design-system/blob/c4f8b395219c701660a11aac922961fe7596bdd4/_scss/all-usajobs.scss#L10)
  * [CivicActions/agile-ms](https://github.com/CivicActions/agile-ms/blob/3093242e2291ed2ea071da26a0a3ecdf81e615c5/docroot/css/_uswds.scss#L11-L12) (see: [_uswds-vars.scss](https://github.com/CivicActions/agile-ms/blob/3093242e2291ed2ea071da26a0a3ecdf81e615c5/docroot/css/_uswds-vars.scss))
  * [FedRamp dashboard](https://github.com/18F/fedramp-dashboard/blob/82393b27dfb6051c607aa7bcd2c3236f392d3c01/src/sass/_variables.scss)
1. Set variables, _then_ import the defaults:
  * [vote.gov Wordpress theme](https://github.com/JHKennedy4/wordpress-theme-votedotorg/blob/a621b3269bd313714541823f1b469860d7b909aa/assets/styles/main.scss#L15-L21)
1. Modify some variables both before _and_ after importing `uswds/all`:
  * [uswds-rails-gem](https://github.com/18F/uswds-rails-gem/blob/master/README.md#usage)

I'm actually kind of relieved that the copy approach isn't the most popular. 😅

**Comment by joshbruce** [comment]
@shawnbot - Looks pretty good! The interesting part of the concept of a `uswds` folder is that it seems to presuppose that customizing will start with a copy of the USWDS project - good on you for searching out the various patterns.

To add a log to the fire: https://github.com/GSA/sam-web-design-standards/tree/gh-pages-staging/src/stylesheets/lib

We copied the project to give us the ability to make a Jekyll site via GitHub. Added the USWDS as a dependency. Updated the build process to copy the `uswds` to `lib`. Then do the component/element-level copy. 

This allows us to override the defaults, without being overwritten upon updates to the USWDS. We can also override (rewrite) the style for individual components/elements. We do the same thing with JS.

**Comment by shawnbot** [comment]
@joshbruce I was actually thinking more of how Bourbon and Neat work, but on further inspection I see that this is only the case when you use the CLI tools to install (see: [Bourbon's instructions](https://github.com/thoughtbot/bourbon/tree/v4-stable#installation) and [Neat's](https://github.com/thoughtbot/neat#using-neat)) a full copy of either library. So I think that means that my original goal in #1417 still stands:

1. If you install uswds via npm or one of the Ruby gems, you should `@import 'uswds';`
1. If you copy the SCSS files into your project directory, you will need to `@import 'path/to/uswds'`

Does that sound right to you?

**Comment by joshbruce** [comment]
@shawnbot - I think so. For me, when I use Bourbon and Neat, I still do the copy to a `vendor` folder, then do the `@import`. So, when I use the USWDS:

1. Add `uswds` to package.json
2. $ npm update
3. $ jekyll build (or whatever build process)
4. Copy `/node_modules/uswds/stylesheets/**/*` to `/assets/vendor/uswds/`
5.`all.scss` contains multiple `@import`

So, it sounds like, I would do something like this - modifying step 4

- Copy `/node_modules/uswds/stylesheets/uswds/**/*` to `/assets/vendor/`

**Comment by shawnbot** [comment]
@joshbruce So currently, you can do this:

```sh
npm install --save uswds
# hopefully this goes into a standalone shell script or a package.json "script"
# so you don't have to remember how to do it when you update to the next version:
rsync -avrC node_modules/uswds/src/stylesheets/ assets/vendor/uswds/
```

then, in your SCSS entrypoint:

```scss
@import 'vendor/uswds/all';
```

The only thing that would change after #1530 merges is your import line:

```scss
@import 'vendor/uswds/uswds';
```

(And this would be optional, because `all.scss` has just been renamed `_all.scss`.)

For Jekyll specifically, I think that we could provide better guidance on simply adding that first path to the config, assuming that the site will always be built in an environment where the npm dependencies are installed:

```yaml
# _config.yml
sass:
  load_paths:
    - node_modules/uswds/src/stylesheets
```

Eventually, I'd love to support npm (and maybe [Eyeglass](http://eyeglass.rocks/)) users more explicitly by providing, in order of roughly increasing complexity:

1. Guidance on adding `node_modules/uswds/src/stylesheets` to your [node-sass `includePaths`](https://github.com/sass/node-sass#includepaths) and importing via `@import 'uswds';`
1. A JS `includePaths` array export that you can use with node-sass's JS API, a la [Bourbon](https://github.com/thoughtbot/bourbon#installing-with-npm-and-using-a-node-based-asset-pipeline) and [Neat](https://github.com/thoughtbot/neat#installing-with-npm-and-using-a-node-based-asset-pipeline). (This is tricky because `require('uswds')` already does something completely different.)
1. An [`eyeglass-exports.js`](https://github.com/sass-eyeglass/eyeglass#in-the-module) that also handles asset path munging.

In the cases of 2 and 3, we (again, _eventually_) wouldn't need to copy the Bourbon and Neat source files because we could include those libraries' include paths in the list we export, which would look something like:

```
/path/to/node_modules/uswds/src/stylesheets
/path/to/node_modules/bourbon/core
/path/to/node_modules/bourbon-neat/app/assets/stylesheets
```

But for now, baby steps! 😉

**Comment by francisrupert** [comment]
@shawnbot,
> "...how people are currently customizing the standards by searching on github and reaching out to those users."

Raising hand! As the Principal UXer at USDA-FSA (by way of SAIC), you can see how I'm doing so at https://github.com/USDA-FSA/fsa-style/blob/master/src/stylesheets/_uswds-select.scss

**[fsa-style](https://github.com/USDA-FSA/fsa-style)** is our starting CSS Framework leaning on your team's awesome work, published as an [NPM module](https://www.npmjs.com/package/fsa-style) and will be documented by the **[fsa-design-system](https://github.com/USDA-FSA/fsa-design-system)**. Baby steps of our own.

**General Summary:**

1. **package.json**: [Install USWDS](https://github.com/USDA-FSA/fsa-style/blob/master/package.json#L38).
1. **Grunt:**: [Copy from `node_modules/uswds/src/` to `src/stylesheets/lib/uswds/`](https://github.com/USDA-FSA/fsa-style/blob/master/Gruntfile.js#L45)
1. **SASS**
  1. [Import select USWDS partials](https://github.com/USDA-FSA/fsa-style/blob/master/src/stylesheets/fsa-style.scss#L3).
  1. [Override / extend variables](https://github.com/USDA-FSA/fsa-style/blob/master/src/stylesheets/fsa-style.scss#L10)
  1. Add in custom `core`, `elements`, `components`, etc.

**Comment by joshbruce** [comment]
@shawnbot & @francisrupert - I think this is where we're going as well. So, it seems like the recommendation is a pretty good solution and I could do something like this:

1. USWDS package.json in project
2. Copy `/node_modules/uswds/stylesheets/uswds/**/*` to `/my/project/stylesheets/lib/uswds`
2. My `all.scss` could `@import 'lib/uswds/all'` to adopt the USWDS wholesale. (Immediate bootstrap.)
3. Then I can iterate as I create my own components, override, and so on. 

Yeah?

**Comment by shawnbot** [comment]
@joshbruce yep, that sounds like what we'll recommend for Jekyll (and other users of non-Node frameworks, static generators, etc.). Note that in the forthcoming release (v0.14.0) you will be able (but not required) to import `lib/uswds/uswds` instead of `lib/uswds/all`.

@francisrupert thanks for weighing in! The same will be true for you, and you may wish to import `lib/uswds/core/variables` instead of `.../core/defaults` (which will be eventually deprecated).

I'm also curious to hear from both of you if you've had to upgrade uswds from npm yet, and what that process was like. If so, did you encounter any issues? I'm mostly wondering if our ([semantic](http://semver.org)) versioning strategy should include Sass partials—e.g. if we rename a partial that any of our users are referencing, maybe we would keep the old partial around and have it `@import` the new one, and clearly document the deprecation of it in the release notes? Maybe there are there internal partials that are essentially "private", and which we can move around without documenting because they're not part of the public "API"?

**Comment by joshbruce** [comment]
@shawnbot - I've updated multiple times, and it's been pretty easy, without much in the way of regression or breakage...except accordions.

As to the other questions. If we are saying `accordions.scss` gets renamed `details.scss` (for [HTML 5.1 `<details>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details)), I think the proposed solution of keeping both around for a bit would be viable. I would hope there would be a sunsetting period though - to keep the clutter minimal (a bunch of Sass files that only import other Sass files could become confusing or intimidating).

As to the private partials notion, I believe we have sort of started this, yeah? The `utilities.scss` could be viewed as the base for something like this. Maybe create a directory inside of `core` called `utilities` and rename `utilities.scss` to `mixins.scss` or something. Though, this concept could get out of hand - tipping the scales in favor of reusability while potentially sacrificing global versus local knowledge. In other words, the learning curve is pretty minimal for the system, because 90% of the information needed for each component can be picked up from looking at the Sass. With a bunch of includes or partials we start losing that ability.

**Comment by shawnbot** [comment]
@joshbruce: Yep! We definitely plan on gradually deprecating the SCSS "API", and even if we were to go the #1526 route (replacing the accordion with `<details>`), I think that we would need to support the "legacy" accordion for at least a full major version. We should explicitly articulate this promise somewhere in our docs! (cc @juliaelman)

We're mulling over a fairly major reorganization of our component styles (no issue yet, as it's not even on the roadmap) that would "colocate" all of each component's assets (SCSS, JS, and potentially HTML template(s)) in their own directory. My hope is that the directory structure provides clearer component boundaries and might allow you to do something like this:

```scss
// in your Sass entry point; imports all the "dependencies" automatically
@import 'uswds/components/header';
```

```js
// in your browserify/webpack/etc. bundle entry point;
// "activates" the header via custom element definition or event delegation
require('uswds/components/header');
```

One thing to consider is that the gradual deprecation approach can be a huge pain to support. I wonder if it's better to be really clear which of our SCSS constitutes a "public API" and state that you `@import` anything else at your own peril? I also started adding [some unit tests](https://github.com/18F/web-design-standards/blob/784bf6c23c419c9f88d586eb8745f3c22ea4b236/spec/sass/include.spec.js) to ensure that both our `uswds` and `all` ("deprecated") entry points work via `@import`; one thing we could do is put tests in place for the most-referenced partials to ensure that upgrades don't break, then remove those tests when we remove those partials for good.

**Comment by joshbruce** [comment]
@shawnbot - Yep. You and I need to find some time to chat in person (or virtual), I think. The reason I say that is because we are currently iterating on something in Angular 2 you might find interesting.

This is not exactly what I'm referring to above.

```
- /core
- /elements
    - /table
        - sass
        - js
        - template
- /components
    - /header
        - ...
- /pages
    - /home
        - ...
_all.scss
```

Re the deprecation pain point - agreed - and maybe putting it as comments in `_all.scss` would be helpful...?? Then, of course, updating the site. Maybe `deprecated` should be part of the maturity scale (if it's not already)...??

**Comment by shawnbot** [comment]
Yes, 💯 to adding "deprecated" to the maturity scale! And yes, let's chat!

**Comment by joshbruce** [comment]
@shawnbot - Added you to a private project...it's been deprecated moving forward, but the Angular one is on an Enterprise GitHub instance, which I don't think I can get you on. Really wish there was a direct message option in GitHub. :)

**Comment by shawnbot** [comment]
This is done in #1536 and #1530! 🎉  @joshbruce and @francisrupert, what do you think of this deprecation notice in the forthcoming release notes?

> 1. We have reorganized some of our SCSS "partials", and will be removing the deprecated paths in a future release. If you import the USWDS into your own Sass or SCSS files, here is what you need to know:
>   * We have replaced the SCSS `all` "entry point" with `uswds`, and set up an alias for `all` (`_all.scss`) to ease the transition for users of previous versions. Upgrade instructions:
>      * **replace** `@import 'all';` with `@import 'uswds';`, or
>      * **replace** `@import 'uswds/all';` to `@import 'uswds/uswds';`
>    * We have deprecated the SCSS `uswds/defaults` partial in favor of `uswds/variables`, and added [`!default`][bang-default] to all of our variables so that they can be selectively overridden by a previous import. Upgrade instructions:
>      * **remove** `@import 'core/defaults';` or `@import 'uswds/core/defaults';`
>      * **add** `@import 'core/variables';` if you don't import it already
>      * **move** your variable customizations before `@import 'core/variables';` for consistency
>      * **remove** any _unmodified_ variable declarations from your customizations to avoid accidentally overriding future changes
>    * `uswds/core/grid-settings` is no longer necessary, as the grid settings are all declared in `uswds/variables`. Upgrade instructions:
>      * **remove** `@import 'core/grid-settings';` or `@import 'uswds/core/grid-settings';`

[bang-default]: https://robots.thoughtbot.com/sass-default
[semver]: http://semver.org/
